apiserver/facades/agent/uniter/uniter.go

```diff
@@ -765,32 +768,45 @@ func (u *UniterAPI) CharmURL(ctx context.Context, args params.Entities) (params.
 			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
 			continue
 		}
-		err = apiservererrors.ErrPerm
-		if canAccess(tag) {
-			var unitOrApplication state.Entity
-			unitOrApplication, err = u.st.FindEntity(tag)
-			if err == nil {
-				var cURL *string
-				var force bool
-
-				switch entity := unitOrApplication.(type) {
-				case *state.Application:
-					cURL, force = entity.CharmURL()
-				case *state.Unit:
-					cURL = entity.CharmURL()
-					// The force value is not actually used on the uniter's unit api.
-					if cURL != nil {
-						force = true
-					}
-				}
+		if !canAccess(tag) {
+			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
+		}
 
-				if cURL != nil {
-					result.Results[i].Result = *cURL
-					result.Results[i].Ok = force
-				}
+		var appName string
+		charmUpgradeOnError := true
+		switch t := tag.(type) {
+		case names.ApplicationTag:
+			appName = t.Id()
+			// The charmUpgradeOnError value is not used by the unit api, so
+			// only set it for applications.
+			charmUpgradeOnError, err = u.applicationService.ShouldAllowCharmUpgradeOnError(ctx, appName)
+			if err != nil {
+				result.Results[i].Error = apiservererrors.ServerError(err)
+				continue
+			}
+		case names.UnitTag:
+			appName, err = names.UnitApplication(t.Id())
+			if err != nil {
+				result.Results[i].Error = apiservererrors.ServerError(err)
+				continue
 			}
+		default:
+			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
 		}
-		result.Results[i].Error = apiservererrors.ServerError(err)
+		charmLocator, err := u.applicationService.GetCharmLocatorByApplicationName(ctx, appName)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		curl, err := apiservercharms.CharmURLFromLocator(charmLocator.Name, charmLocator)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		result.Results[i].Result = curl
+		result.Results[i].Ok = charmUpgradeOnError
```

Comment:
This is impossible to read. Can we create a method that handles, just the application and just the unit.